### ISSUE TITLE
LoadPicture extensions

### ISSUE THREAD

**Issue by KarlSevenSeven** [twinbasic/twinbasic#1282]
**Is your feature request related to a problem? Please describe.**

VB LoadPicture is not so smart.
In tB some extensions would be nice:

- PNG support
- Load from resource
- Load from external resource DLL

When it is possible to load PNG's, then it would be good to be able to save them to disk as well.

In VB it's a lot of effort to load/save PNG without dependencies.

**Describe the solution you'd like**
LoadPicture/SavePicture accepting PNG

**Describe alternatives you've considered**
WIA or other means

**Comment by bclothier** [comment]
I would like to add that it actually may be preferable to support vector-based images. This would make it much easier to build applications that can handle different DPIs/monitor resolutions/size/whatever without pixellating.

**Comment by WaynePhillipsEA** [comment]
Just FYI, SVG support would be great, but it's a very difficult format to support fully.  There is no decent library that I've seen that has a suitable licence for it to be integrated into tB. 

Recent versions of Windows do offer some built-in support for SVGs, but if we go down that route we'd be limiting the feature to some edition of Win 10 onwards (I don't recall which edition of Win 10 it finally came supported with).

PNG support, on the other would be an easy low-hanger as we already use LodePNG inside of tB IDE.

**Comment by KarlSevenSeven** [comment]
Regarding icons etc., SVG promises more than it can deliver in my experience.
When icons are small, the rendering is seldom acceptable.
IMO it's better to have a set of icons in different sizes, e.g. 24/32/48...
And yes, that's effort.

**Comment by jpbro** [comment]
> Just FYI, SVG support would be great, but it's a very difficult format to support fully. There is no decent library that I've seen that has a suitable licence for it to be integrated into tB.

This library is MPLv2: https://github.com/RazrFalcon/resvg
Here's a Simplified BSD-licensed VB6 wrapper for resvg by @tannerhelland:  https://github.com/tannerhelland/vb6-svg

AFAIK both Simplified BSD and MPLv2 are very permissive open source licenses, though I'm not sure if they fit your particular needs.

**Comment by WaynePhillipsEA** [comment]
@jpbro AFAIU, it's not ideal for us because of the copyleft licencing.  With us being a commercial non-open source project, we'd have to ship it as a separate DLL, but that pretty much goes against the "portable" nature of tB builds.   Thanks though, I wasn't aware of that one.

**Comment by jpbro** [comment]
I'm not a lawyer, so take this with a grain of salt, but as far as I can tell from the [MPLv2 FAQ](https://www.mozilla.org/en-US/MPL/2.0/FAQ/), you can compile/statically link MPLv2 code into an EXE. See the last sentence of the answer for Q11:

> **Q11: How 'viral' is the MPL? If I use MPL-licensed code in my proprietary application, will I have to give all the source code away?**
> No. The license requires that Modifications (as defined in Section 1.10 of the license) must be licensed under the MPL and made available to anyone to whom you distribute the Source Code. However, new files containing no MPL-licensed code are not Modifications, and therefore do not need to be distributed under the terms of the MPL, even if you create a Larger Work (as defined in Section 1.7) by using, compiling, or distributing the non-MPL files together with MPL-licensed files. **_This allows, for example, programs using MPL-licensed code to be statically linked to and distributed as part of a larger proprietary piece of software, which would not generally be possible under the terms of stronger copyleft licenses._**

You would still need to make the MPLv2 portions of the source code available. If they're unmodified, it may be enough to link back to the original project (though I'm not sure about this). However, because devs will be compiling and distributing their own software using tB, this might still be undesirable as it would place a burden on everyone who distributes tB compiled code that they might not want to deal with.

**Comment by rexxitall** [comment]
How about killing 1000 lawyers - usually we call it a good start :D JOKE !

**Comment by WaynePhillipsEA** [comment]
@jpbro Interesting!  Thanks.   Certainly at first glance, that does appear to be correct.  I will look into this some more. 

Perhaps it could be a per-project opt-in switch to enable SVG support, with a warning _(reminder)_ outputted during the tB build process that you need to link back to the main repo (or a fork of ours) as per MPLv2 licence requirements.

**Comment by bclothier** [comment]
I think a more sustainable solution would be to make it a package that leverages IDE's extensibility API for design-time support bits.

**Comment by WaynePhillipsEA** [comment]
So, going back to PNG support.  Do we need some sort of switch to enable/disable such extensions? (e.g an optional flags parameter).  Technically it's a change of behaviour.

**Comment by Kr00l** [comment]
What picture type will the OlePicture have from a PNG? Is it a 32 bit bitmap or a 24 or 32 bit icon ?

**Comment by WaynePhillipsEA** [comment]
I guess bitmap would be the expected type, but that could be a choice, again via flags potentially.

**Comment by Kr00l** [comment]
I didn' test but maybe returning same picture type as following code would do is a good start:
```
With New WIA.ImageFile
        .LoadFile "Misc1.png"
        Set Image1.Picture = .FileData.Picture
    End With
```

What would be the breaker of supporting PNG? And what scenario would it make necessary to turn it off?
One could still resort to stdole.LoadPicture() to get the VB6 behavior. (No bytestream etc; which tB already supports which could be a "breaker" seen also.)

**Comment by WaynePhillipsEA** [comment]
True... perhaps I'm being over cautious.  We could cross that bridge when we come to it, if anyone complains ;)

**Comment by fafalone** [comment]
I'm not sure I understand how this would break compatibility; what VB6 code might fail if this is supported in tB?

**Comment by WaynePhillipsEA** [comment]
It's more that it is a difference of behaviour.  The old LoadPicture will throw an error for PNGs. The new one won't.   Perhaps someone catches that error in VBx and deals with PNGs differently.  It's not a likely problem... just a possible, unlikely scenario.  Like I said, i was probably overthinking it.

**Comment by WaynePhillipsEA** [comment]
`Global.LoadPicture` now supports PNG files as of BETA 159.   Note that the OLE provided version of `LoadPicture` does not.
